crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,492 @@
+            Console::new(0, vec![0; MAX_CAPACITY]).unwrap_err(),
+            Error::InvalidMaximumCapacity
+        );
+
+        let console = Console::new(MAX_CAPACITY, vec![0; MAX_CAPACITY]).unwrap();
+        assert_eq!(
+            console
+                .enqueue_data(&mut vec![INPUT_VALUE; MAX_CAPACITY])
+                .unwrap(),
+            ()
+        );
+        assert_eq!(
+            console.enqueue_data(&mut vec![INPUT_VALUE; 1]).unwrap_err(),
+            Error::BufferCapacityExceeded
+        );
+
+        console.clear_input_buffer();
+        assert_eq!(console.is_input_buffer_empty(), true);
+    }
+
+    #[test]
+    fn test_process_transmitq_chain() {
+        let output: Vec<u8> = Vec::new();
+        let mut console = Console::new(MAX_CAPACITY, output).unwrap();
+
+        let mem: GuestMemoryMmap =
+            GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 0x10_000)]).unwrap();
+
+        // One descriptor is write only
+        let v = vec![
+            Descriptor::new(0x1000, INPUT_SIZE as u32, 0, 0),
+            Descriptor::new(0x2000, INPUT_SIZE as u32, VIRTQ_DESC_F_WRITE, 0),
+        ];
+
+        let queue = MockSplitQueue::new(&mem, 16);
+        // The `build_desc_chain` function will populate the `NEXT` related flags and field.
+        let mut chain = queue.build_desc_chain(&v[..2]).unwrap();
+        assert_eq!(
+            console.process_transmitq_chain(&mut chain).unwrap_err(),
+            Error::UnexpectedWriteOnlyDescriptor
+        );
+
+        // Descriptor is outside of the memory bounds
+        let v = vec![
+            Descriptor::new(0x10_000, INPUT_SIZE as u32, 0, 0),
+            Descriptor::new(0x20_000, INPUT_SIZE as u32, 0, 0),
+        ];
+        let mut chain = queue.build_desc_chain(&v[..2]).unwrap();
+        assert_eq!(
+            console.process_transmitq_chain(&mut chain).unwrap_err(),
+            Error::WriteToOutputFailed(GuestMemoryError::InvalidGuestAddress(GuestAddress(
+                0x10_000
+            )))
+        );
+
+        // Test normal functionality.
+        let v = vec![
+            Descriptor::new(0x3000, INPUT_SIZE as u32, 0, 0),
+            Descriptor::new(0x4000, INPUT_SIZE as u32, 0, 0),
+        ];
+        let mut chain = queue.build_desc_chain(&v[..2]).unwrap();
+        mem.write_slice(&vec![INPUT_VALUE; INPUT_SIZE], GuestAddress(0x3000))
+            .unwrap();
+        mem.write_slice(&vec![INPUT_VALUE * 2; INPUT_SIZE], GuestAddress(0x4000))
+            .unwrap();
+        console.output.clear();
+        assert_eq!(
+            console.process_transmitq_chain(&mut chain).unwrap(),
+            2 * INPUT_SIZE
+        );
+        assert_eq!(console.output[..INPUT_SIZE], vec![INPUT_VALUE; INPUT_SIZE]);
+        assert_eq!(
+            console.output[INPUT_SIZE..],
+            vec![INPUT_VALUE * 2; INPUT_SIZE]
+        );
+    }
+
+    #[test]
+    fn test_process_receiveq_chain() {
+        let guest_mem = &mut [0u8; INPUT_SIZE];
```

Comment:
Why do we call this `guest_mem`?

Reply:
I used this buffer to copy slices of `GuestMemory` and compare it with the input data.